nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/SameDiffLoss.java

```diff
@@ -0,0 +1,146 @@
+package org.nd4j.linalg.lossfunctions;
+
+import org.nd4j.autodiff.samediff.SDVariable;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.base.Preconditions;
+import org.nd4j.linalg.activations.IActivation;
+import org.nd4j.linalg.api.buffer.DataType;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.primitives.Pair;
+
+import java.util.Collections;
+import java.util.Map;
+
+
+public abstract class SameDiffLoss implements ILossFunction {
+
+    SameDiff sd = SameDiff.create();
+
+
+    protected SameDiffLoss() {
+
+
+        SDVariable layerInput =  sd.placeHolder("layerInput", DataType.FLOAT ,-1);
+        SDVariable labels = sd.placeHolder("labels", DataType.FLOAT ,-1);
+        this.defineLoss(sd, layerInput, labels);
+
+
+    }
+
+
+    public abstract SDVariable defineLoss(SameDiff sd, SDVariable layerInput, SDVariable labels);
+
+    /**
+     * Compute the score (loss function value) for the given inputs.
+     * @param labels       Label/expected preOutput
+     * @param preOutput    Output of the model (neural network)
+     * @param activationFn Activation function that should be applied to preOutput
+     * @param mask         Mask array; may be null
+     * @param average      Whether the score should be averaged (divided by number of rows in labels/preOutput) or not   @return Loss function value
+     */
+    public double computeScore(INDArray labels, INDArray preOutput, IActivation activationFn, INDArray mask, boolean average) {
+
+            // The score overall consists of the
+            // sum of the negative log likelihoods for each
+            // of the individual labels.
+
+
+
+        INDArray scoreArr = computeScoreArray(labels, preOutput, activationFn, mask);
+
+        double score = scoreArr.sumNumber().doubleValue();
+            if (average) {
+                score /= scoreArr.size(0);
+            }
+            return score;
+        }
+
+
+
+
+
+
+    /**
+     * Compute the score (loss function value) for each example individually.
+     * For input [numExamples,nOut] returns scores as a column vector: [numExamples,1]
+     * @param labels       Labels/expected output
+     * @param preOutput    Output of the model (neural network)
+     * @param activationFn Activation function that should be applied to preOutput
+     * @param mask         @return Loss function value for each example; column vector
+     */
+    public INDArray computeScoreArray(INDArray labels, INDArray preOutput, IActivation activationFn, INDArray mask){
+
+        Preconditions.checkArgument((labels.size(1) != preOutput.size(1)),"Labels array numColumns (size(1) = " + labels.size(1) + ") does not match output layer\n" +
+                "number of outputs (nOut = " + preOutput.size(1) + ")");
```

Comment:
When using the Preconditions class, do this instead:
This avoids object creation/garbage unless an error is actually thrown.
```
Preconditions.checkArgument((labels.size(1) != preOutput.size(1)),"Labels array numColumns (size(1) = %s) does not match output layer number of outputs (nOut = %s)",  labels.size(1), preOutput.size(1));
```